src/mpi/romio/adio/common/ad_close.c

```diff
@@ -97,13 +97,8 @@ void ADIO_Close(ADIO_File fd, int *error_code)
     MPI_Comm_free(&(fd->comm));
     ADIOI_Free(fd->filename);
 
-    MPI_Type_get_envelope(fd->etype, &i, &j, &k, &combiner);
-    if (combiner != MPI_COMBINER_NAMED)
-        MPI_Type_free(&(fd->etype));
-
-    MPI_Type_get_envelope(fd->filetype, &i, &j, &k, &combiner);
-    if (combiner != MPI_COMBINER_NAMED)
-        MPI_Type_free(&(fd->filetype));
+    ADIOI_Type_dispose(&(fd->etype));
+    ADIOI_Type_dispose(&(fd->filetype));
 
```

Comment:
good cleanup.  ROMIO dates back to the "function calls are expensive" era, so there are lots of opportunities to add back in some utilities.  thanks.

Reply:
I am testing an ideal of exposing MPICH typerep functionality as MPIX functions, in particular, two -
* `MPIX_Type_iov(datatype, iov_offset, iov[], max_iov_len, *actual_iov_len)`
* `MPIX_Type_iov_len(datatype, max_iov_bytes, *iov_len)`

With these two, user can query and retrieve contiguous segments at any offset. Thus ROMIO can skip maintaining the flat list. The flat list code are not only complex, it can also be very slow and potentially use up the memory.

Reply:
> ROMIO dates back to the "function calls are expensive" era

Good old times... Now kiddos code Python for a living!